packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -1045,35 +1022,53 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
 
             switch (summaryObject.type) {
                 case api.SummaryType.Tree: {
+                    treesPathToSummaryTreeCacheLatest.set(`.app${path}/${key}`, summaryObject);
                     const result = await this.convertSummaryToSnapshotTree(
                         parentHandle,
                         summaryObject,
                         blobsShaToPathCacheLatest,
+                        blobsPathToShaCacheLatest,
+                        treesPathToSummaryTreeCacheLatest,
                         depth + 1,
-                        `${path}/${key}`);
+                        `${path}/${key}`,
+                        expanded);
                     value = result.snapshotTree;
                     reusedBlobs += result.reusedBlobs;
                     blobs += result.blobs;
                     break;
                 }
                 case api.SummaryType.Blob: {
-                    value = typeof summaryObject.content === "string"
+                    let hash: string | undefined;
+                    let completePath: string | undefined;
+                    const blobPath = `.app${path}/${key}`;
+                    // If we are expanding the handle, then the blobPath should exist in the cache as we will get the blob
+                    // hash from the cache.
+                    if (expanded) {
+                        hash = this.blobsPathToShaCache.get(blobPath);
+                        assert(hash !== undefined, "hash should be set here");
+                        completePath = this.blobsShaToPathCache.get(hash);
+                        assert(completePath !== undefined, "path should be defined as path->sha mapping exists");
+                    } else {
+                        value = typeof summaryObject.content === "string"
                         ? { content: summaryObject.content, encoding: "utf-8" }
                         : { content: Uint8ArrayToString(summaryObject.content, "base64"), encoding: "base64" };
+                        hash = await hashFile(IsoBuffer.from(value.content, value.encoding));
+                    }
 
-                    // Promises for pending hashes in blobsCachePendingHashes should all have resolved and removed themselves
-                    assert(this.blobsCachePendingHashes.size === 0);
-                    const hash = await hashFile(IsoBuffer.from(value.content, value.encoding));
-                    let completePath = this.blobsShaToPathCache.get(hash);
+                    assert(hash !== undefined, "hash should be set!");
                     // If the cache has the hash of the blob and handle of last summary is also present, then use that
                     // to generate complete path for the given blob.
                     if (!completePath || !this.lastSummaryHandle) {
```

Comment:
I'm a bit puzzled by this IF condition. Part of me wants to say that's equivalent to if (!expanded).
The only place where it seems no true is for the first summary, as we set this.lastSummaryHandle only on writing summary, and not have it when loading from snapshot. Which begs the question - do we do wrong things on first summary? Why should it be different? I believe it should not, but I'm not sure if we have a handle from loading from summary.

Same question regarding this.blobsShaProposalHandle and this check in uploadSummaryWithContext - are we not doing any reuse on first summary?
        if (context.proposalHandle !== this.blobsShaProposalHandle) {

Reply:
"lastSummaryHandle" should never be undefined for detach new because summarizer client always load from a summary for detach new. However I am not assuming that detach new is only case until we remove old code support from container.load(). 
"completePath" or "cachedPath" could be found even when expanded = false. So this is not equal to (!expanded),